### PR TITLE
Correct statistics (filter pruning and sparsity)

### PR DESCRIPTION
**Pull request by negvet**
### Changes

Statistics of the pruned model are now collected from the model (using get_nncf_operations) but not from the graph (affected when loading from a checkpoint, the graph is not getting updated, but the model is).

When loading from a checkpoint the current_level of `Scheduler` is recalculated.

`current_step` and `current_epoch` attributes of BaseCompressionScheduler become private (just getters are available - not possible to update from outside). Tests are updated in accordance (no direct set of `current_step` and `current_epoch`).

### Reason for changes

Printed statistics of the model and algorithm are not correct while loading from the checkpoint.

### Tests
Updated.

### CODE REVIEW

**Comment by negvet** [comment]
Jenkins please retry a build

File: nncf/tensorflow/pruning/base_algorithm.py [hunk 1]
```diff
@@ -300,6 +300,31 @@ def _check_pruning_level(self, params):
...
+                        mask = wrapped_layer.ops_weights[op_name]['mask']
+                        reduce_axes = list(range(len(mask.shape)))
+                        filter_axis = get_filter_axis(wrapped_layer, weight_attr)
+                        if filter_axis == -1:
+                            filter_axis = reduce_axes[filter_axis]
+                        reduce_axes.remove(filter_axis)
+                        filter_mask = tf.reduce_max(tf.cast(mask, tf.int32), axis=reduce_axes, keepdims=True)
+                        filters_number = get_filters_num(wrapped_layer)
+                        pruned_filter_number = filters_number - tf.reduce_sum(filter_mask)
+                        num_of_sparse_elements_by_node[wrapped_layer.name] = pruned_filter_number.numpy()
```
Comment (daniil-lyakhov) [inline nncf/tensorflow/pruning/base_algorithm.py:319]: Please consider to unify this part with `_calculate_pruned_layers_summary`

File: nncf/common/pruning/utils.py [hunk 2]
```diff
@@ -385,8 +384,7 @@ def get_num_of_sparse_elements_by_node(node_name: str) -> int:
 
 
 def calculate_in_out_channels_by_masks(pruning_groups: List[Cluster[PrunedLayerInfoBase]],
-                                       masks: Dict[str, NNCFTensor],
-                                       tensor_processor: Type[NNCFCollectorTensorProcessor],
+                                       num_of_sparse_elements_by_node: Dict[str, int],
```
Comment (vshampor) [inline nncf/common/pruning/utils.py:387]: Should use `NNCFNodeName` instead of `str`

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/common/sparsity/schedulers.py [hunk 3]
```diff
@@ -89,6 +89,17 @@ def current_sparsity_level(self) -> float:
...
         return self._current_level
 
+    def load_state(self, state: Dict[str, Any]) -> None:
+        super().load_state(state)
+        if self._current_level is None:
+            self._current_level = self._calculate_sparsity_level()
+
+    def get_state(self) -> Dict[str, Any]:
+        if self._current_level is None:
+            self._current_level = self._calculate_sparsity_level()
```
Comment (andrey-churkin) [inline nncf/common/sparsity/schedulers.py:99]: For which purposes is it needed here?

File: nncf/common/pruning/schedulers.py [hunk 4]
```diff
@@ -108,6 +109,11 @@ def current_pruning_level(self) -> float:
             return self._current_level
         return 0
 
+    def load_state(self, state: Dict[str, Any]) -> None:
+        super().load_state(state)
+        if self._current_level is None:
+            self._current_level = self._calculate_pruning_level()
```
Comment (daniil-lyakhov) [inline nncf/common/pruning/schedulers.py:115]: Please consider implementing `get_state` to save `_current_level` and modifying loading from given state, otherwise `self_current_level` always will be `None` on checkpoint loading. 

We should decide should we keep current pruning level is state dict or we should always calculate it (in this case if statement above should be removed)

File: nncf/common/sparsity/schedulers.py [hunk 5]
```diff
@@ -89,6 +89,17 @@ def current_sparsity_level(self) -> float:
...
+        state = super().get_state()
+        return state
```
Comment (daniil-lyakhov) [inline nncf/common/sparsity/schedulers.py:101]: Same comment about `current_level` parameter saving to state

File: nncf/common/sparsity/schedulers.py [hunk 6]
```diff
@@ -277,7 +291,7 @@ def __init__(self, controller: SparsityController, params: dict):
         self.schedule = MultiStepSchedule(
             sorted(params.get('multistep_steps', [90])), params.get('multistep_sparsity_levels', [0.1, 0.5]))
         self.target_level = self.schedule.values[-1]
-        self._current_level = self.schedule.values[0]
+        self._current_level = None
```
Comment (andrey-churkin) [inline nncf/common/sparsity/schedulers.py:294]: I think you can remove this line because the same is in the `SparsityScheduler` superclass.

File: nncf/tensorflow/pruning/base_algorithm.py [hunk 7]
```diff
@@ -10,7 +10,7 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 """
-
+from typing import Dict
```
Comment (andrey-churkin) [inline nncf/tensorflow/pruning/base_algorithm.py:13]: Could you please add a blank line before the line `from typing import Dict`?

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (alexsu52) [reply nncf/common/pruning/schedulers.py:115]: It is preferable to calculate than to store. @daniil-lyakhov, what do you mean under otherwise `self_current_level` always will be `None` on checkpoint loading? I see the opposite:
```
        if self._current_level is None:
            self._current_level = self._calculate_pruning_level()
```

File: nncf/tensorflow/pruning/base_algorithm.py [hunk 8]
```diff
@@ -300,6 +300,31 @@ def _check_pruning_level(self, params):
         if pruning_flops_target:
             self.prune_flops = True
 
+    def _calculate_num_of_sparse_elements_by_node(self) -> Dict[str, int]:
+        num_of_sparse_elements_by_node = {}
+        wrapped_layers = collect_wrapped_layers(self._model)
+        for wrapped_layer in wrapped_layers:
+            for weight_attr, ops in wrapped_layer.weights_attr_ops.items():
+                for op_name in ops:
+                    if op_name in self._op_names:
```
Comment (andrey-churkin) [inline nncf/tensorflow/pruning/base_algorithm.py:309]: I think you can use `get_nncf_operations()` method from the `nncf.tensorflow.graph.utils` module. Could you please look at this?

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (daniil-lyakhov) [reply nncf/common/pruning/schedulers.py:115]: I mean, if  `load_state` always will be called before `epoch_step`, `_current_level`  after scheduler initialization always will be `None` in `load_state`. In case one would load model after `epoch_step` call, all scheduler stats will be updated besides `_current_level`, which will be properly recalculated only on `epoch_step`

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (andrey-churkin) [reply nncf/common/pruning/schedulers.py:115]: I don't understand the problem.

> I mean, if `load_state` always will be called before `epoch_step`, `_current_level` after scheduler initialization always will be `None` in `load_state`.

Seems correct. If `_current_level` is `None` inside `load_state()` it will be calculated according to `current_epoch` and `current_step` attributes.

> In case one would load model after `epoch_step` call, all scheduler stats will be updated besides `_current_level`, which will be properly recalculated only on `epoch_step`

Seems like it is the wrong case. The model should be initialized before `epoch_step()`.

File: nncf/common/pruning/utils.py [hunk 9]
```diff
@@ -396,16 +394,14 @@ def calculate_in_out_channels_by_masks(pruning_groups: List[Cluster[PrunedLayerI
...
+    :param num_of_sparse_elements_by_node: A dictionary of num_of_sparse_elements of each pruning node.
     :param full_input_channels:  A dictionary of input channels number in original model.
     :param full_output_channels: A dictionary of output channels number in original model.
     :param pruning_groups_next_nodes: A dictionary of next nodes of each pruning group.
     :return Dictionary of new input channels number {node_name: channels_num}
     """
     def get_num_of_sparse_elements_by_node(node_name: str) -> int:
-        mask = masks[node_name]
-        return mask.shape[0] - int(tensor_processor.sum(mask))
+        return num_of_sparse_elements_by_node[node_name]
```
Comment (evgeniya-egupova) [inline nncf/common/pruning/utils.py:404]: You can pass the `num_of_sparse_elements_by_node` dictionary to `_calculate_in_out_channels` as is.

Comment (evgeniya-egupova) [reply nncf/common/pruning/utils.py:404]: And move code from `_calculate_in_out_channels` to `calculate_in_out_channels_by_masks` and rename it to `calculate_in_out_channels`.

File: nncf/common/sparsity/schedulers.py (hunk 3 above)
Comment (negvet) [reply nncf/common/sparsity/schedulers.py:99]: We have a test (not sure if it is a valid one, but probably this case can be present sometimes).
https://github.com/openvinotoolkit/nncf/blob/8bcf5d243e53afca9511457dffbd7f6048669ba9/tests/torch/sparsity/test_common.py#L93-L122
After creating the scheduler (`self._current_level` is still `None`), the parameters of the scheduler are changed manually and  `get_compression_state()` is called.
```
ctrl.scheduler.current_step = 100
ctrl.scheduler.current_epoch = 5
compression_state = ctrl.get_compression_state()
```
To be precise here, the scheduler should recalculate the sparsity level in accordance with the current epoch and step.

File: nncf/tensorflow/pruning/base_algorithm.py (hunk 1 above)
Comment (negvet) [reply nncf/tensorflow/pruning/base_algorithm.py]: Thanks for the comment! Unified.

File: nncf/common/pruning/utils.py (hunk 2 above)
Comment (negvet) [reply nncf/common/pruning/utils.py]: Thanks for the comment! Updated.

File: nncf/common/sparsity/schedulers.py (hunk 6 above)
Comment (negvet) [reply nncf/common/sparsity/schedulers.py]: Thanks for the comment! Updated

File: nncf/tensorflow/pruning/base_algorithm.py (hunk 7 above)
Comment (negvet) [reply nncf/tensorflow/pruning/base_algorithm.py:14]: Thanks for the comment! Updated

File: nncf/tensorflow/pruning/base_algorithm.py (hunk 8 above)
Comment (negvet) [reply nncf/tensorflow/pruning/base_algorithm.py]: Thanks for the suggestion. Indeed I can, used.

File: nncf/tensorflow/pruning/base_algorithm.py [hunk 10]
```diff
@@ -300,6 +300,31 @@ def _check_pruning_level(self, params):
         if pruning_flops_target:
             self.prune_flops = True
 
+    def _calculate_num_of_sparse_elements_by_node(self) -> Dict[str, int]:
```
[the remark this answers is not in the document]
Comment (negvet) [reply nncf/tensorflow/pruning/base_algorithm.py]: Thanks for the comment! Added

File: nncf/tensorflow/sparsity/callbacks.py [hunk 11]
```diff
@@ -47,9 +47,11 @@ def _prepare_for_tensorboard(self, stats: NNCFStatistics):
         tensorboard_stats = {
             f'{base_prefix}/sparsity_level_for_model': ms.sparsity_level,
             f'{base_prefix}/sparsity_level_for_sparsified_layers': ms.sparsity_level_for_layers,
-            f'{base_prefix}/target_sparsity_level': stats.target_sparsity_level,
         }
 
+        if stats.target_sparsity_level is not None:
+            tensorboard_stats.update({f'{base_prefix}/target_sparsity_level': stats.target_sparsity_level})
```
[the remark this answers is not in the document]
Comment (negvet) [reply nncf/tensorflow/sparsity/callbacks.py]: Thanks for the comment! Updated

File: nncf/common/pruning/utils.py (hunk 9 above)
Comment (negvet) [reply nncf/common/pruning/utils.py:404]: There is another method `calculate_in_out_channels_in_uniformly_pruned_model` which is using `_calculate_in_out_channels` with its custom sparse element counter.

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (daniil-lyakhov) [reply nncf/common/pruning/schedulers.py:116]: @andrey-churkin, I know this isn't real case, but if one start training from scratch and then, after training, load checkpoint, the model contains `_current_level` from scratch training

Comment (daniil-lyakhov) [reply nncf/common/pruning/schedulers.py:116]: I suggest always recalculate `_current_level` on `load_model`

File: nncf/tensorflow/pruning/base_algorithm.py (hunk 1 above)
Comment (daniil-lyakhov) [reply nncf/tensorflow/pruning/base_algorithm.py]: 👍

**Review by daniil-lyakhov — APPROVED** [review]

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (daniil-lyakhov) [reply nncf/common/pruning/schedulers.py]: `_current_level` attr was removed as discussed

**Comment by daniil-lyakhov** [comment]
Jenkins please retry a build

File: nncf/common/sparsity/schedulers.py [hunk 12]
```diff
@@ -74,20 +74,21 @@ def _update_sparsity_level(self) -> None:
...
         """
-        self._current_level = self._calculate_sparsity_level()
         if self.current_epoch >= self.freeze_epoch:
             self._controller.freeze()
-        self._controller.set_sparsity_level(self._current_level)
+        self._controller.set_sparsity_level(self._calculate_sparsity_level())
 
     @property
-    def current_sparsity_level(self) -> float:
+    def current_sparsity_level(self) -> Union[float, None]:
```
Comment (vshampor) [inline nncf/common/sparsity/schedulers.py:82]: `Optional[float]`. Should also document in the docstring specifically the situations when `None` may be returned.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/common/sparsity/schedulers.py (hunk 12 above)
Comment (negvet) [reply nncf/common/sparsity/schedulers.py]: Thanks for the comment. Fixed!

**Comment by negvet** [comment]
Jenkins please retry a build

**Comment by negvet** [comment]
Jenkins please retry a build

**Comment by negvet** [comment]
Jenkins please retry a build

**Comment by negvet** [comment]
Jenkins please retry a build

File: nncf/torch/sparsity/magnitude/algo.py [hunk 13]
```diff
@@ -158,7 +158,7 @@ def compression_stage(self) -> CompressionStage:
         if self._mode == 'local':
             return CompressionStage.FULLY_COMPRESSED
 
-        if self.scheduler.current_sparsity_level == 0:
+        if self.scheduler.current_sparsity_level is None or self.scheduler.current_sparsity_level == 0:
```
Comment (alexsu52) [inline nncf/torch/sparsity/magnitude/algo.py:161]: It is a very difficult criterion for UNCOMPRESSED compression stage. I would recommend return None or create separate method for it.

File: tests/tensorflow/sparsity/magnitude/test_algorithm.py [hunk 14]
```diff
@@ -71,10 +71,8 @@ def test_compression_controller_state():
     _, compression_ctrl = create_compressed_model_and_algo_for_test(model, config)
 
     # Test get state
-    compression_ctrl.scheduler.current_step = 100
-    compression_ctrl.scheduler.current_epoch = 5
```
Comment (alexsu52) [inline tests/tensorflow/sparsity/magnitude/test_algorithm.py:75]: Why did not you set the step and epoch for the scheduler?

File: tests/tensorflow/sparsity/test_common.py [hunk 15]
```diff
@@ -86,9 +86,9 @@ def test_can_create_sparse_scheduler__with_defaults(self, algo):
         assert scheduler.freeze_epoch == 100
 
     @pytest.mark.parametrize(('schedule', 'get_params', 'ref_levels'),
-                             (('polynomial', get_poly_params, [0.2, 0.2, 0.4, 0.6, 0.6, 0.6, 0.6]),
-                              ('exponential', get_poly_params, [0.2, 0.2, 0.4343145, 0.6, 0.6, 0.6, 0.6]),
-                              ('multistep', get_multistep_params, [0.2, 0.2, 0.2, 0.4, 0.5, 0.6, 0.6])))
+                             (('polynomial', get_poly_params, [0.2, 0.4, 0.6, 0.6, 0.6, 0.6]),
+                              ('exponential', get_poly_params, [0.2, 0.4343145, 0.6, 0.6, 0.6, 0.6]),
+                              ('multistep', get_multistep_params, [0.2, 0.2, 0.4, 0.5, 0.6, 0.6])))
```
Comment (alexsu52) [inline tests/tensorflow/sparsity/test_common.py:91]: What is the reason for these changes?

File: nncf/common/sparsity/schedulers.py [hunk 16]
```diff
@@ -232,6 +232,17 @@ def __init__(self, controller: SparsityController, params: dict):
         self.eps = params.get('eps', 0.03)
         self.patience = params.get('patience', 1)
         self.num_bad_epochs = 0
+        self._current_level = self.initial_level
+
+    @property
+    def current_sparsity_level(self) -> Optional[float]:
```
Comment (andrey-churkin) [inline nncf/common/sparsity/schedulers.py:238]:
```suggestion
    def current_sparsity_level(self) -> float:
```

File: nncf/common/sparsity/schedulers.py [hunk 17]
```diff
@@ -277,7 +290,18 @@ def __init__(self, controller: SparsityController, params: dict):
         self.schedule = MultiStepSchedule(
             sorted(params.get('multistep_steps', [90])), params.get('multistep_sparsity_levels', [0.1, 0.5]))
         self.target_level = self.schedule.values[-1]
-        self._current_level = self.schedule.values[0]
+
+    @property
+    def current_sparsity_level(self) -> Optional[float]:
```
Comment (andrey-churkin) [inline nncf/common/sparsity/schedulers.py:295]:
```suggestion
    def current_sparsity_level(self) -> float:
```

File: nncf/tensorflow/sparsity/callbacks.py [hunk 18]
```diff
@@ -47,7 +47,6 @@ def _prepare_for_tensorboard(self, stats: NNCFStatistics):
         tensorboard_stats = {
             f'{base_prefix}/sparsity_level_for_model': ms.sparsity_level,
             f'{base_prefix}/sparsity_level_for_sparsified_layers': ms.sparsity_level_for_layers,
-            f'{base_prefix}/target_sparsity_level': stats.target_sparsity_level,
```
Comment (andrey-churkin) [inline nncf/tensorflow/sparsity/callbacks.py:50]: What is the reason for this change?

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

File: nncf/tensorflow/sparsity/callbacks.py (hunk 18 above)
Comment (negvet) [reply nncf/tensorflow/sparsity/callbacks.py:50]: Thanks for the notice. It is a leftover. Fixed.

File: nncf/common/sparsity/schedulers.py (hunk 17 above)
Comment (negvet) [reply nncf/common/sparsity/schedulers.py:295]: Updated, thanks!

File: nncf/common/sparsity/schedulers.py (hunk 16 above)
Comment (negvet) [reply nncf/common/sparsity/schedulers.py:238]: Updated, thanks!

File: nncf/common/sparsity/schedulers.py [hunk 19]
```diff
@@ -74,20 +73,21 @@ def _update_sparsity_level(self) -> None:
...
         """
-        self._current_level = self._calculate_sparsity_level()
         if self.current_epoch >= self.freeze_epoch:
             self._controller.freeze()
-        self._controller.set_sparsity_level(self._current_level)
+        self._controller.set_sparsity_level(self._calculate_sparsity_level())
 
     @property
-    def current_sparsity_level(self) -> float:
+    def current_sparsity_level(self) -> Optional[float]:
```
[the remark this answers is not in the document]
Comment (negvet) [reply nncf/common/sparsity/schedulers.py:81]: Updated, thanks!

**Review by andrey-churkin — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]

File: nncf/common/pruning/utils.py [hunk 20]
```diff
@@ -397,16 +395,14 @@ def calculate_in_out_channels_by_masks(pruning_groups: List[Cluster[PrunedLayerI
...
     :param pruning_groups: A list of pruning groups.
-    :param masks: A dictionary of masks of each pruning node.
-    :param tensor_processor: NNCF Tensor processor to operate on NNCFTensors.
+    :param num_of_sparse_elements_by_node: A dictionary of num_of_sparse_elements of each pruning node.
     :param full_input_channels:  A dictionary of input channels number in original model.
     :param full_output_channels: A dictionary of output channels number in original model.
     :param pruning_groups_next_nodes: A dictionary of next nodes of each pruning group.
     :return Dictionary of new input channels number {node_name: channels_num}
     """
     def get_num_of_sparse_elements_by_node(node_name: str) -> int:
```
[the remark this answers is not in the document]
Comment (negvet) [reply nncf/common/pruning/utils.py:404]: Thanks for the notice.
`_calculate_in_out_channels` expects `get_num_of_sparse_elements_by_node` to be callable. Please refer to `calculate_in_out_channels_in_uniformly_pruned_model` method as well.
